### DISCUSSION TITLE
How to turn off compression

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#2967] [Q&A] [answered]
### Question

# Question

How can I disable compression of http data when running nicegui?  Profiler shows my nicegui app spends a lot of time compressing data.  It's not needed and contributing to lag.

There's no obvious flag in the [ui.run page](https://nicegui.io/documentation/run), and the [uvicorn.run args](https://www.uvicorn.org/deployment/) aren't clear (ws-per-message-deflate is the only option that mentions compression, but this doesn't seem like the right setting).  Can't find anything relevant in issue discussions either.

**Comment by rodja** [comment, marked as answer]
You are right. Currently the gzip compression middleware is non-optional because it's simply added when importing NiceGUI:

https://github.com/zauberzeug/nicegui/blob/3fcfac637273ebcd051cd09c1ee5d7745863d044/nicegui/nicegui.py#L57

We should move it to the `ui_run.py` initialization and introduce a configuration parameter. Would you like to create a pull request? Until it is merged you may be able to access `app.user_middleware` and remove the middleware in your application code...

**Reply by ed2050** [reply]
@rodja Ok thanks.  I'll see what I can do.  The [fastapi](https://fastapi.tiangolo.com/advanced/middleware/#gzipmiddleware) and [starlette](https://www.starlette.io/middleware/) middleware pages don't mention removing middleware handlers.  Not sure it's possible once they're added.

> Would you like to create a pull request?

I would _like_ to, the question is whether I'm _able_ to 😅.  We'll see...

**Reply by me21** [reply]
Here's how I did it in one old application for NiceGUI 0.9.x:

```python
if __name__ == '__main__':
    # monkey patch JustPy to remove GZipMiddleware (causes high CPU usage during database backup file download)
    app: Starlette = nicegui.globals.app
    app.user_middleware = [m for m in app.user_middleware if m.cls is not starlette.middleware.gzip.GZipMiddleware]
    app.middleware_stack = app.build_middleware_stack()
```

Hope it will work with NiceGUI 1.4.x!

**Reply by ed2050** [reply]
Thanks @me21!  Seems to do the trick.  I'm on nicegui 1.3.16.

**Comment by sklibanov** [comment]
Looks like there's a new parameter to ui.run():

        gzip_middleware_factory: Callable[[ASGIApp], GZipMiddleware] | None = GZipMiddleware,

from here - https://github.com/zauberzeug/nicegui/pull/5582 .

Good! Thanks!
